File: 7.4.pl
#!/usr/bin/perl

use strict;
use warnings;

use MyLinkedList;

package MyLLAreJoined;

sub new {
    return bless {} => shift;
}

sub ll_are_joined{
    my ($self, $ll1, $ll2) = @_;
    my ($big_iter1, $big_iter2, $count) = ($ll1->{first}, $ll2->{first}, 1);
    while (defined($big_iter1->{next_node}) || defined($big_iter2->{next_node})) { # FIXME we don't test last item then!!!
        my ($i, $iter) = ($count, $ll1->{first});
        while (defined($iter->{next_node}) && $i > 0) {
            if ($big_iter2 == $iter) {
                return $iter;
            }
            $iter = $iter->{next_node};
            $i--;
        }
        ($i, $iter) = ($count, $ll2->{first});
        while (defined($iter->{next_node}) && $i > 0) {
            if ($big_iter1 == $iter) {
                return $iter;
            }
            $iter = $iter->{next_node};
            $i--;
        }
        $big_iter1 = $big_iter1->{next_node};
        $big_iter2 = $big_iter2->{next_node};
        $count++;
    }
    if ($big_iter1 == $big_iter2) { return $big_iter1 }
    return undef;
}

package TestMyLLAreJoined;
use Test::More;
use Test::Deep;

my $module = MyLLAreJoined->new;

my ($ll1, $ll2) = (MyLinkedList->new(), MyLinkedList->new());
is($module->ll_are_joined($ll1, $ll2), undef, 'undef for empty');

($ll1, $ll2) = (MyLinkedList->new(1), MyLinkedList->new());
is($module->ll_are_joined($ll1, $ll2), undef, 'one is empty');

($ll1, $ll2) = (MyLinkedList->new(), MyLinkedList->new(2,3,4));
is($module->ll_are_joined($ll1, $ll2), undef, 'first is empty');

($ll1, $ll2) = (MyLinkedList->new(1,2,3), MyLinkedList->new(4,5));
use Data::Dumper;
is($module->ll_are_joined($ll1, $ll2), undef, "not joined anyway if I haven't joined them=)");

$ll2->{first}->{next_node} = $ll1->{first}->{next_node};
#warn Dumper($ll1);
#warn Dumper($ll2);
is_deeply($module->ll_are_joined($ll1, $ll2), {value => 2, next_node => { value => 3, next_node => undef }}, 'joined now!');

($ll1, $ll2) = (MyLinkedList->new(1,2), MyLinkedList->new(4,5));
$ll2->{first}->{next_node} = $ll1->{first}->{next_node};
is_deeply($module->ll_are_joined($ll1, $ll2), {value => 2, next_node => undef}, 'short, joined');

($ll1, $ll2) = (MyLinkedList->new(1,2,3,4), MyLinkedList->new(5,6,7));
$ll1->{first}->{next_node}->{next_node} = $ll2->{first};
is_deeply($module->ll_are_joined($ll1, $ll2), {value => 5, next_node => { value => 6, next_node => { value => 7, next_node => undef }}}, 'joined at the start');

done_testing;

File: 7.4.py
#!/usr/bin/env python

from mynode import *
import unittest

def ll_are_joined(ll1, ll2):
    if ll1 is None or ll2 is None:
        return None
    len1 = length(ll1) # count lengths
    len2 = length(ll2)
    if len1 > len2:
        ll1, ll2 = ll2, ll1
        len1, len2 = len2, len1
    # now ll1 is the shorter list
    while len2 > len1:
        ll2 = ll2.next_node
        len2 -= 1
    while ll1 is not ll2:
        ll1, ll2 = ll1.next_node, ll2.next_node
    return ll1

def length(ll):
    length = 0
    while ll is not None:
        length += 1
        ll = ll.next_node
    return length


class MyTest(unittest.TestCase):

    def test_small(self):
        first = ll_are_joined(from_array([]), from_array([]))
        self.assertEqual(first, None, 'empty')

        first = ll_are_joined(from_array([1]), from_array([]))
        self.assertEqual(first, None, 'still empty')

        first = ll_are_joined(from_array([]), from_array([1]))
        self.assertEqual(first, None, 'still empty again')

    def test_not_joined(self):
        first = ll_are_joined(from_array([2]), from_array([3]))
        self.assertEqual(first, None, 'one and one, not joined')

        first = ll_are_joined(from_array([2,3,4,5]), from_array([2,3,4,5]))
        self.assertEqual(first, None, 'not joined still')

    def test_joined(self):
        ll1, ll2 = from_array([1,3]), from_array([2,3])
        ll1.next_node = ll2.next_node
        first = ll_are_joined(ll1, ll2)
        self.assertEqual(first.value, 3, 'same len, last elem')

        ll1, ll2 = from_array([1,3,4]), from_array([2,3,5,6,7])
        ll1.next_node = ll2.next_node
        first = ll_are_joined(ll1, ll2)
        self.assertEqual(first.value, 3, 'same len, not last elem')

        ll1, ll2 = from_array([1,2,3,4]), from_array([4,5,6])
        ll1.next_node.next_node = ll2
        first = ll_are_joined(ll1, ll2)
        self.assertEqual(first.value, 4, 'joined at the start of second list')

        ll1, ll2 = from_array([1,2]), from_array([4,5,6])
        ll1.next_node.next_node = ll2.next_node
        first = ll_are_joined(ll1, ll2)
        self.assertEqual(first.value, 5, 'joined at the end of first list')


if __name__ == '__main__':
    suite = unittest.TestLoader().loadTestsFromTestCase(MyTest)
    unittest.TextTestRunner(verbosity=2).run(suite)
